packages/iov-crypto/src/libsodium.spec.ts

```diff
@@ -621,4 +621,204 @@ describe("Libsodium", () => {
+        await Xchacha20poly1305Ietf.encrypt(message, key, nonce)
+          .then(() => fail("encryption must not succeed"))
+          .catch(error => expect(error).toMatch(/invalid key length/));
+      }
+      {
+        // 64 bytes
+        const key = fromHex("1324cdddc4b94e625bbabcac862c9429ba011e2184a1ccad60e7c3f6ff4916d81324cdddc4b94e625bbabcac862c9429ba011e2184a1ccad60e7c3f6ff4916d8") as Xchacha20poly1305IetfKey;
+        await Xchacha20poly1305Ietf.encrypt(message, key, nonce)
+          .then(() => fail("encryption must not succeed"))
+          .catch(error => expect(error).toMatch(/invalid key length/));
+      }
+    });
+
+    it("decryption fails with wrong ciphertext/key/nonce", async () => {
+      const key = fromHex("1324cdddc4b94e625bbabcac862c9429ba011e2184a1ccad60e7c3f6ff4916d8") as Xchacha20poly1305IetfKey;
+      const nonce = fromHex("000000000000000000000000000000000000000000000000") as Xchacha20poly1305IetfNonce;
+
+      const originalMessage = new Uint8Array([0x11, 0x22, 0x33, 0x44]) as Xchacha20poly1305IetfMessage;
+      const ciphertext = await Xchacha20poly1305Ietf.encrypt(originalMessage, key, nonce);
+      expect(ciphertext).toBeTruthy();
+      expect(ciphertext.length).toBeTruthy(4 /* message length */ + 32 /* tag length*/);
+
+      {
+        // baseline
+        expect(await Xchacha20poly1305Ietf.decrypt(ciphertext, key, nonce)).toEqual(originalMessage);
+      }
+      {
+        // corrupted ciphertext
+        const corruptedCiphertext = ciphertext.map((x, i) => (i === 0 ? x ^ 0x01 : x)) as Xchacha20poly1305IetfCiphertext;
+        await Xchacha20poly1305Ietf.decrypt(corruptedCiphertext, key, nonce)
+          .then(() => fail("promise must not resolve"))
+          .catch(error => expect(error.message).toContain("invalid usage"));
+      }
+      {
+        // corrupted key
+        const corruptedKey = key.map((x, i) => (i === 0 ? x ^ 0x01 : x)) as Xchacha20poly1305IetfKey;
+        await Xchacha20poly1305Ietf.decrypt(ciphertext, corruptedKey, nonce)
+          .then(() => fail("promise must not resolve"))
+          .catch(error => expect(error.message).toContain("invalid usage"));
+      }
+      {
+        // corrupted nonce
+        const corruptedNonce = nonce.map((x, i) => (i === 0 ? x ^ 0x01 : x)) as Xchacha20poly1305IetfNonce;
+        await Xchacha20poly1305Ietf.decrypt(ciphertext, key, corruptedNonce)
+          .then(() => fail("promise must not resolve"))
+          .catch(error => expect(error.message).toContain("invalid usage"));
+      }
+    });
+
+    it("encrypt conforms to Botan implementation ", async () => {
+      // Test data generated by
+      // echo -n "<message>" | ./botan encryption --mode=chacha20poly1305 --iv=000000000000000000000000000000000000000000000000 --ad= --key=0000000000000000000000000000000000000000000000000000000000000000 | xxd -p -c 1000000
```

Comment:
Nice test cases.
Can you provide a link to where I can install `./botan`?

Also, why is it always with 000 for iv and key? Seems that it doesn't cover all cases?
Nevermind... Just saw that you do have random values for iv and key below. 👍

Reply:
Support for Xchacha is only available on master right now and not yet released.

```
$ git clone https://github.com/randombit/botan.git
$ cd botan
$ ./configure.py && make -j 4
$ ./botan --help
```

I created 100k test vectors to check the Botan implementation is okay ;) https://github.com/webmaster128/xchacha20poly1305-testvectors

Reply:
Awesome. I know it all works properly, just suggesting a code comment to refer to this, if anyone comes in later to do an audit.